### PR TITLE
Add base system boilerplate

### LINKED ISSUE #1: Missing base system boilerplate

### PR DESCRIPTION
**Pull request by jcmfernandes**
Closes #1

### CODE REVIEW

File: config/logger.rb [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    end
+
+    private
+
+    def format_hash(hash, keys = nil)
+      kvs = (keys || hash.keys).map do |k|
+        "#{k}=\"#{hash[k]}\"" if hash.has_key?(k)
+      end
+      str = kvs.tap(&:compact!).join(' ')
+      str.empty? ? nil : str
```
Comment (ivoanjo) [inline config/logger.rb]: This can just be `str unless str.empty?`

File: lib/bairrobot/error.rb [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    def self.create(title: nil)
+      return Class.new(self) if title.nil?
+
+      klass = Class.new(self) do
+        def title
+          self.class.const_get(:TITLE, false)
+        end
+      end
+
+      klass.const_set(:TITLE, title.dup.freeze)
```
Comment (ivoanjo) [inline lib/bairrobot/error.rb]: I'm 95% sure you can just do `TITLE = tile.dup.freeze` inside the `Class.new` block above.

File: lib/bairrobot/error.rb [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    attr_reader :context
+    attr_reader :id
+    attr_writer :wrapped_exception
+
+    def self.create(title: nil)
+      return Class.new(self) if title.nil?
+
+      klass = Class.new(self) do
+        def title
+          self.class.const_get(:TITLE, false)
```
Comment (ivoanjo) [inline lib/bairrobot/error.rb]: I believe you can just do `def title; TITLE; end` here?

File: spec/spec_helper.rb [hunk 4]
```diff
@@ -0,0 +1,68 @@
+ENV['RACK_ENV'] = 'none'.freeze
```
Comment (ivoanjo) [inline spec/spec_helper.rb]: Maybe `test` would be clearer here?

File: spec/spec_helper.rb [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+ENV['APP_ENV']  = 'test'.freeze
```
Comment (ivoanjo) [inline spec/spec_helper.rb]: Maybe using the frozen string pragma would be easier here?

File: Gemfile [hunk 6]
```diff
@@ -0,0 +1,52 @@
+source 'https://rubygems.org'
+
+ruby IO.read(File.join(__dir__, '.ruby-version')).strip
```
Comment (jcmfernandes) [inline Gemfile]: The value is inside file `.ruby-version`. This is just a way to make everyone happy as some platforms like to look at the Gemfile for the ruby version, some others prefer the `.ruby-version` file, so this unifies that.

File: Gemfile [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+gem 'faraday', '~> 0.12.1'
+gem 'excon', '~> 0.57.0'
+
+# Logging
+gem 'logstash-logger', '~> 0.25.1'
+gem 'liquid', '~> 4.0'
+
+# Data serialization
+gem 'oj', '~> 3.0'
+gem 'multi_json', '~> 1.12', '>= 1.12.1'
```
Comment (jcmfernandes) [inline Gemfile]: Lets leave it as it's something I'm absolutely sure we will need.

File: Gemfile [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+gem 'dry-auto_inject', '< 0.5', '>= 0.4.2'
+gem 'dry-configurable', '< 0.8', '>= 0.7'
+gem 'dry-equalizer', '< 0.3', '>= 0.2'
+gem 'dry-transaction', '< 0.11', '>= 0.10.2'
+gem 'dry-monads', '< 0.4', '>= 0.3.1'
+gem 'dry-validation', '< 0.11', '>= 0.10.6'
+gem 'dry-struct', '< 0.4', '>= 0.3'
+
+# Hash utilities
+gem 'activesupport', '~> 5.0.0'
```
Comment (jcmfernandes) [inline Gemfile]: Yes we need it, using it for the `Error` class. And yes, we can update it, will do that! 😉

File: lib/bairrobot/env.rb [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+  app_env == 'staging'.freeze
+end
+
+def development_env?
+  app_env == 'development'.freeze
+end
+
+def test_env?
+  app_env == 'test'.freeze
+end
```
Comment (jcmfernandes) [inline lib/bairrobot/env.rb]: NO!

File: lib/bairrobot/env.rb [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+def set_env(env)
+  return unless block_given?
+
+  CURRENT_ENVIRONMENT.bind(env.to_s.dup.freeze) { yield }
+end
```
Comment (jcmfernandes) [inline lib/bairrobot/env.rb]: Will do.

File: config/logger.rb (hunk 1 above)
Comment (jcmfernandes) [inline config/logger.rb]: Yes it can!

File: config/boot.rb [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+
+require_relative 'environment'
+require_relative 'connections'
+require_relative 'dependencies'
+
+Bairrobot.logger.progname = app_name
+$LOGGER = $LOG = Bairrobot.logger
+
+Dir.glob(File.join(__dir__, 'initializers', '*.rb')).each do |file|
+  load file
```
Comment (jcmfernandes) [inline config/boot.rb]: Likely no good reason. I think I had problems requiring files this way in the past and ended up going for load, knowing that I won't be requiring/loading these files elsewhere. It was probably a stupid mistake at the time, I'll check.

File: lib/bairrobot/error.rb (hunk 3 above)
Comment (jcmfernandes) [inline lib/bairrobot/error.rb]: As explained IRL, it can't.

File: lib/bairrobot/error.rb (hunk 2 above)
Comment (jcmfernandes) [inline lib/bairrobot/error.rb]: Again, as explained IRL, nope, you can't.

File: spec/spec_helper.rb (hunk 5 above)
Comment (jcmfernandes) [inline spec/spec_helper.rb]: Indeed, will do that 😉 .

File: spec/spec_helper.rb (hunk 4 above)
Comment (jcmfernandes) [inline spec/spec_helper.rb]: Well, `test` means nothing to rack. Valid values are `development`, `deployment`, and `none`. You check their semantics [here](https://github.com/rack/rack/blob/5c36acac4d298dcf04c58a10014c849a478da53f/lib/rack/server.rb#L158).

File: Gemfile (hunk 7 above)
Comment (jcmfernandes) [inline Gemfile]: I'm actually already using it in the logger's boiler plate to output in logstash format (JSON).

**Pull request merged**